{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3619ab59",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:59:16.586880Z",
     "start_time": "2021-07-12T10:59:16.582725Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import jpy_tools.parseSnake2 as jps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6d2de13b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:59:17.361872Z",
     "start_time": "2021-07-12T10:59:17.117100Z"
    }
   },
   "outputs": [],
   "source": [
    "path_snakeFile = '/public/home/liuzj/scripts/pipeline/10xAnalysis_concatenatedGenome/snakemake/snakefile'\n",
    "path_config = '/public/home/liuzj/scripts/pipeline/10xAnalysis_concatenatedGenome/snakemake/config.yaml'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d4a5ef93",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:59:17.845920Z",
     "start_time": "2021-07-12T10:59:17.841899Z"
    }
   },
   "outputs": [],
   "source": [
    "sf = jps.SnakeFile()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "16c5574f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:59:20.040849Z",
     "start_time": "2021-07-12T10:59:19.822706Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/home/liuzj/softwares/anaconda3/lib/python3.8/site-packages/jpy_tools/parseSnake2.py:93: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  self.yaml = yaml.load(open(self.path))\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "import pandas as pd\n",
       "#configfile: \"/public/home/liuzj/scripts/pipeline/10xAnalysis_concatenatedGenome/snakemake/config.yaml\"\n",
       "pipelineDir = config['pipelineDir']\n",
       "resultDir = config[\"resultDir\"].rstrip(\"/\") + \"/\"\n",
       "pipelineDir = config[\"pipelineDir\"].rstrip(\"/\") + \"/\"\n",
       "sampleLs = list(config['samples'].keys())"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header = jps.SnakeHeader(sf, path_config)\n",
    "header.addCode(\"sampleLs = list(config['samples'].keys())\")\n",
    "config = header.getConfig()\n",
    "header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a8554b5f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T11:01:40.642785Z",
     "start_time": "2021-07-12T11:01:40.620600Z"
    }
   },
   "outputs": [],
   "source": [
    "import glob\n",
    "df_runStarsolo = pd.DataFrame(config['samples']).T\n",
    "df_runStarsolo = df_runStarsolo.assign(starIndex = config['concatenatedStarIndex'], outputDir=df_runStarsolo.index + '/')\n",
    "df_runStarsolo['sampleR2'] = df_runStarsolo['sampleDir'].map(lambda x:','.join(glob.glob(f\"{x}/*R2*.fastq.gz\")))\n",
    "df_runStarsolo['sampleR1'] = df_runStarsolo['sampleDir'].map(lambda x:','.join(glob.glob(f\"{x}/*R1*.fastq.gz\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2fa2e349",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:57:36.965724Z",
     "start_time": "2021-07-12T10:57:36.952104Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-12 18:57:36.956 | INFO     | jpy_tools.parseSnake2:addRule:55 - runStarsolo step num: 1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "import glob\n",
       "df_runStarsolo = pd.DataFrame(config['samples']).T\n",
       "df_runStarsolo = df_runStarsolo.assign(starIndex = config['concatenatedStarIndex'], outputDir=df_runStarsolo.index + '/')\n",
       "df_runStarsolo['sampleR2'] = df_runStarsolo['sampleDir'].map(lambda x:','.join(glob.glob(f\"{x}/*/*R2*.fastq.gz\")))\n",
       "df_runStarsolo['sampleR1'] = df_runStarsolo['sampleDir'].map(lambda x:','.join(glob.glob(f\"{x}/*/*R1*.fastq.gz\")))\n",
       "for column in ['outputDir']:\n",
       "    df_runStarsolo[column] = resultDir + 'step1_runStarsolo/' + df_runStarsolo[column]\n",
       "----------------\n",
       "IN RULE\n",
       "----------------\n",
       "rule runStarsolo:\n",
       "    input:\n",
       "        sampleDir = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'sampleDir'],\n",
       "    output:\n",
       "        runStarsoloFinished = resultDir + 'step1_runStarsolo/' + '{sample}.finished',\n",
       "    params:\n",
       "        gpu = 0,\n",
       "        expectedCell = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'expectedCell'],\n",
       "        umiLength = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'umiLength'],\n",
       "        whiteList = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'whiteList'],\n",
       "        includeIntron = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'includeIntron'],\n",
       "        starIndex = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'starIndex'],\n",
       "        outputDir = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'outputDir'],\n",
       "        sampleR2 = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'sampleR2'],\n",
       "        sampleR1 = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'sampleR1'],\n",
       "    threads:32\n",
       "    priority:20\n",
       "    shell:\n",
       "        \"\"\"\n",
       "mkdir -p {params.outputDir}\n",
       "cd {params.outputDir}\n",
       "if [ {params.includeIntron} = True ]\n",
       "then\n",
       "    STAR --runThreadN {threads} --genomeDir {params.starIndex} --soloType CB_UMI_Simple --soloCBwhitelist {params.whiteList}        --soloUMIlen {params.umiLength} --readFilesIn {params.sampleR2} {params.sampleR1} --soloFeatures Gene GeneFull Velocyto        --clipAdapterType CellRanger4 --outFilterScoreMin 30 --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --soloUMIfiltering MultiGeneUMI_CR        --soloUMIdedup 1MM_CR --soloCellFilter  EmptyDrops_CR {params.expectedCell} 0.99 10 45000 90000 500 0.01 20000 0.01 10000         --readFilesCommand zcat\n",
       "else\n",
       "    STAR --runThreadN {threads} --genomeDir {params.starIndex} --soloType CB_UMI_Simple --soloCBwhitelist {params.whiteList}        --soloUMIlen {params.umiLength} --readFilesIn {params.sampleR2} {params.sampleR1} --soloFeatures Gene Velocyto        --clipAdapterType CellRanger4 --outFilterScoreMin 30 --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --soloUMIfiltering MultiGeneUMI_CR        --soloUMIdedup 1MM_CR --soloCellFilter  EmptyDrops_CR {params.expectedCell} 0.99 10 45000 90000 500 0.01 20000 0.01 10000         --readFilesCommand zcat\n",
       "fi\n",
       "samtools view -@ {threads} -S -O BAM ./Aligned.out.sam  | samtools sort -@ {threads} - -O BAM -o ./Aligned.out.bam\n",
       "samtools index -@ {threads} ./Aligned.out.bam\n",
       "rm ./Aligned.out.sam\n",
       "touch {output.runStarsoloFinished}\n",
       "        \"\"\""
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "runStarsolo = jps.SnakeRule(sf, \"runStarsolo\", 32, priority=20)\n",
    "runStarsolo.addCode(\n",
    "    \"\"\"\n",
    "import glob\n",
    "df_runStarsolo = pd.DataFrame(config['samples']).T\n",
    "df_runStarsolo = df_runStarsolo.assign(starIndex = config['concatenatedStarIndex'], outputDir=df_runStarsolo.index + '/')\n",
    "df_runStarsolo['sampleR2'] = df_runStarsolo['sampleDir'].map(lambda x:','.join(glob.glob(f\"{x}/*R2*.fastq.gz\")))\n",
    "df_runStarsolo['sampleR1'] = df_runStarsolo['sampleDir'].map(lambda x:','.join(glob.glob(f\"{x}/*R1*.fastq.gz\")))\n",
    "\"\"\"\n",
    ")\n",
    "runStarsolo.addMetaDf(\"df_runStarsolo\", [\"outputDir\"])\n",
    "runStarsolo.addMain(\"input\", [\"sampleDir\"])\n",
    "runStarsolo.addMain(\n",
    "    \"params\",\n",
    "    [\n",
    "        \"expectedCell\",\n",
    "        \"umiLength\",\n",
    "        \"whiteList\",\n",
    "        \"includeIntron\",\n",
    "        \"starIndex\",\n",
    "        \"outputDir\",\n",
    "        \"sampleR2\",\n",
    "        \"sampleR1\",\n",
    "    ],\n",
    ")\n",
    "runStarsolo.setShell(\n",
    "    \"\"\"\n",
    "mkdir -p {params.outputDir}\n",
    "cd {params.outputDir}\n",
    "if [ {params.includeIntron} = True ]\n",
    "then\n",
    "    STAR --runThreadN {threads} --genomeDir {params.starIndex} --soloType CB_UMI_Simple --soloCBwhitelist {params.whiteList}\\\n",
    "        --soloUMIlen {params.umiLength} --readFilesIn {params.sampleR2} {params.sampleR1} --soloFeatures Gene GeneFull Velocyto\\\n",
    "        --clipAdapterType CellRanger4 --outFilterScoreMin 30 --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --soloUMIfiltering MultiGeneUMI_CR\\\n",
    "        --soloUMIdedup 1MM_CR --soloCellFilter  EmptyDrops_CR {params.expectedCell} 0.99 10 45000 90000 500 0.01 20000 0.01 10000 \\\n",
    "        --readFilesCommand zcat\n",
    "else\n",
    "    STAR --runThreadN {threads} --genomeDir {params.starIndex} --soloType CB_UMI_Simple --soloCBwhitelist {params.whiteList}\\\n",
    "        --soloUMIlen {params.umiLength} --readFilesIn {params.sampleR2} {params.sampleR1} --soloFeatures Gene Velocyto\\\n",
    "        --clipAdapterType CellRanger4 --outFilterScoreMin 30 --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --soloUMIfiltering MultiGeneUMI_CR\\\n",
    "        --soloUMIdedup 1MM_CR --soloCellFilter  EmptyDrops_CR {params.expectedCell} 0.99 10 45000 90000 500 0.01 20000 0.01 10000 \\\n",
    "        --readFilesCommand zcat\n",
    "fi\n",
    "samtools view -@ {threads} -S -O BAM ./Aligned.out.sam  | samtools sort -@ {threads} - -O BAM -o ./Aligned.out.bam\n",
    "samtools index -@ {threads} ./Aligned.out.bam\n",
    "rm ./Aligned.out.sam\n",
    "\"\"\"\n",
    ")\n",
    "runStarsolo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "80c78541",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:57:36.982482Z",
     "start_time": "2021-07-12T10:57:36.967895Z"
    }
   },
   "outputs": [],
   "source": [
    "df_runCellRanger = pd.DataFrame(config['samples']).T\n",
    "\n",
    "df_runCellRanger = df_runCellRanger.explode(\"runCellRanger\")\n",
    "df_runCellRanger.index = df_runCellRanger.index + \"_\" + df_runCellRanger.runCellRanger\n",
    "df_runCellRanger['cellRangerIndex'] = df_runCellRanger['runCellRanger'].map(config['dt_genome_cellRangerIndex'])\n",
    "df_runCellRanger['id'] = df_runCellRanger.index\n",
    "df_runCellRanger['outDir'] = df_runCellRanger.index\n",
    "sample_specieLs = df_runCellRanger.index.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f6cc08cc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:57:37.030909Z",
     "start_time": "2021-07-12T10:57:36.984654Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>runCellRanger</th>\n",
       "      <th>sampleDir</th>\n",
       "      <th>expectedCell</th>\n",
       "      <th>umiLength</th>\n",
       "      <th>whiteList</th>\n",
       "      <th>includeIntron</th>\n",
       "      <th>r2Length</th>\n",
       "      <th>cellRangerIndex</th>\n",
       "      <th>id</th>\n",
       "      <th>outDir</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>anteriorCingulateCortex_v2_hg</th>\n",
       "      <td>hg</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/hg38_enb/</td>\n",
       "      <td>anteriorCingulateCortex_v2_hg</td>\n",
       "      <td>anteriorCingulateCortex_v2_hg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>anteriorCingulateCortex_v2_panpan</th>\n",
       "      <td>panpan</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/panpan/</td>\n",
       "      <td>anteriorCingulateCortex_v2_panpan</td>\n",
       "      <td>anteriorCingulateCortex_v2_panpan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>anteriorCingulateCortex_v2_panTro</th>\n",
       "      <td>panTro</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/panTro/</td>\n",
       "      <td>anteriorCingulateCortex_v2_panTro</td>\n",
       "      <td>anteriorCingulateCortex_v2_panTro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>anteriorCingulateCortex_v2_mmul</th>\n",
       "      <td>mmul</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/mmul_10/</td>\n",
       "      <td>anteriorCingulateCortex_v2_mmul</td>\n",
       "      <td>anteriorCingulateCortex_v2_mmul</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>caudateNucleus_v2_hg</th>\n",
       "      <td>hg</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/hg38_enb/</td>\n",
       "      <td>caudateNucleus_v2_hg</td>\n",
       "      <td>caudateNucleus_v2_hg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>caudateNucleus_v2_panpan</th>\n",
       "      <td>panpan</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/panpan/</td>\n",
       "      <td>caudateNucleus_v2_panpan</td>\n",
       "      <td>caudateNucleus_v2_panpan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>caudateNucleus_v2_panTro</th>\n",
       "      <td>panTro</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/panTro/</td>\n",
       "      <td>caudateNucleus_v2_panTro</td>\n",
       "      <td>caudateNucleus_v2_panTro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>caudateNucleus_v2_mmul</th>\n",
       "      <td>mmul</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/mmul_10/</td>\n",
       "      <td>caudateNucleus_v2_mmul</td>\n",
       "      <td>caudateNucleus_v2_mmul</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cerebellarGrayMatter_v2_hg</th>\n",
       "      <td>hg</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/hg38_enb/</td>\n",
       "      <td>cerebellarGrayMatter_v2_hg</td>\n",
       "      <td>cerebellarGrayMatter_v2_hg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cerebellarGrayMatter_v2_panpan</th>\n",
       "      <td>panpan</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/panpan/</td>\n",
       "      <td>cerebellarGrayMatter_v2_panpan</td>\n",
       "      <td>cerebellarGrayMatter_v2_panpan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cerebellarGrayMatter_v2_panTro</th>\n",
       "      <td>panTro</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/panTro/</td>\n",
       "      <td>cerebellarGrayMatter_v2_panTro</td>\n",
       "      <td>cerebellarGrayMatter_v2_panTro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cerebellarGrayMatter_v2_mmul</th>\n",
       "      <td>mmul</td>\n",
       "      <td>/public/home/liuzj/projects/singleCell/primate...</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>/public/home/liuzj/data/10xWhiteList/10xV2.txt</td>\n",
       "      <td>True</td>\n",
       "      <td>98</td>\n",
       "      <td>/public/home/liuzj/data/singleCellRef/mmul_10/</td>\n",
       "      <td>cerebellarGrayMatter_v2_mmul</td>\n",
       "      <td>cerebellarGrayMatter_v2_mmul</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  runCellRanger  \\\n",
       "anteriorCingulateCortex_v2_hg                hg   \n",
       "anteriorCingulateCortex_v2_panpan        panpan   \n",
       "anteriorCingulateCortex_v2_panTro        panTro   \n",
       "anteriorCingulateCortex_v2_mmul            mmul   \n",
       "caudateNucleus_v2_hg                         hg   \n",
       "caudateNucleus_v2_panpan                 panpan   \n",
       "caudateNucleus_v2_panTro                 panTro   \n",
       "caudateNucleus_v2_mmul                     mmul   \n",
       "cerebellarGrayMatter_v2_hg                   hg   \n",
       "cerebellarGrayMatter_v2_panpan           panpan   \n",
       "cerebellarGrayMatter_v2_panTro           panTro   \n",
       "cerebellarGrayMatter_v2_mmul               mmul   \n",
       "\n",
       "                                                                           sampleDir  \\\n",
       "anteriorCingulateCortex_v2_hg      /public/home/liuzj/projects/singleCell/primate...   \n",
       "anteriorCingulateCortex_v2_panpan  /public/home/liuzj/projects/singleCell/primate...   \n",
       "anteriorCingulateCortex_v2_panTro  /public/home/liuzj/projects/singleCell/primate...   \n",
       "anteriorCingulateCortex_v2_mmul    /public/home/liuzj/projects/singleCell/primate...   \n",
       "caudateNucleus_v2_hg               /public/home/liuzj/projects/singleCell/primate...   \n",
       "caudateNucleus_v2_panpan           /public/home/liuzj/projects/singleCell/primate...   \n",
       "caudateNucleus_v2_panTro           /public/home/liuzj/projects/singleCell/primate...   \n",
       "caudateNucleus_v2_mmul             /public/home/liuzj/projects/singleCell/primate...   \n",
       "cerebellarGrayMatter_v2_hg         /public/home/liuzj/projects/singleCell/primate...   \n",
       "cerebellarGrayMatter_v2_panpan     /public/home/liuzj/projects/singleCell/primate...   \n",
       "cerebellarGrayMatter_v2_panTro     /public/home/liuzj/projects/singleCell/primate...   \n",
       "cerebellarGrayMatter_v2_mmul       /public/home/liuzj/projects/singleCell/primate...   \n",
       "\n",
       "                                  expectedCell umiLength  \\\n",
       "anteriorCingulateCortex_v2_hg            10000        10   \n",
       "anteriorCingulateCortex_v2_panpan        10000        10   \n",
       "anteriorCingulateCortex_v2_panTro        10000        10   \n",
       "anteriorCingulateCortex_v2_mmul          10000        10   \n",
       "caudateNucleus_v2_hg                     10000        10   \n",
       "caudateNucleus_v2_panpan                 10000        10   \n",
       "caudateNucleus_v2_panTro                 10000        10   \n",
       "caudateNucleus_v2_mmul                   10000        10   \n",
       "cerebellarGrayMatter_v2_hg               10000        10   \n",
       "cerebellarGrayMatter_v2_panpan           10000        10   \n",
       "cerebellarGrayMatter_v2_panTro           10000        10   \n",
       "cerebellarGrayMatter_v2_mmul             10000        10   \n",
       "\n",
       "                                                                        whiteList  \\\n",
       "anteriorCingulateCortex_v2_hg      /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "anteriorCingulateCortex_v2_panpan  /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "anteriorCingulateCortex_v2_panTro  /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "anteriorCingulateCortex_v2_mmul    /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "caudateNucleus_v2_hg               /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "caudateNucleus_v2_panpan           /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "caudateNucleus_v2_panTro           /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "caudateNucleus_v2_mmul             /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "cerebellarGrayMatter_v2_hg         /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "cerebellarGrayMatter_v2_panpan     /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "cerebellarGrayMatter_v2_panTro     /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "cerebellarGrayMatter_v2_mmul       /public/home/liuzj/data/10xWhiteList/10xV2.txt   \n",
       "\n",
       "                                  includeIntron r2Length  \\\n",
       "anteriorCingulateCortex_v2_hg              True       98   \n",
       "anteriorCingulateCortex_v2_panpan          True       98   \n",
       "anteriorCingulateCortex_v2_panTro          True       98   \n",
       "anteriorCingulateCortex_v2_mmul            True       98   \n",
       "caudateNucleus_v2_hg                       True       98   \n",
       "caudateNucleus_v2_panpan                   True       98   \n",
       "caudateNucleus_v2_panTro                   True       98   \n",
       "caudateNucleus_v2_mmul                     True       98   \n",
       "cerebellarGrayMatter_v2_hg                 True       98   \n",
       "cerebellarGrayMatter_v2_panpan             True       98   \n",
       "cerebellarGrayMatter_v2_panTro             True       98   \n",
       "cerebellarGrayMatter_v2_mmul               True       98   \n",
       "\n",
       "                                                                   cellRangerIndex  \\\n",
       "anteriorCingulateCortex_v2_hg      /public/home/liuzj/data/singleCellRef/hg38_enb/   \n",
       "anteriorCingulateCortex_v2_panpan    /public/home/liuzj/data/singleCellRef/panpan/   \n",
       "anteriorCingulateCortex_v2_panTro    /public/home/liuzj/data/singleCellRef/panTro/   \n",
       "anteriorCingulateCortex_v2_mmul     /public/home/liuzj/data/singleCellRef/mmul_10/   \n",
       "caudateNucleus_v2_hg               /public/home/liuzj/data/singleCellRef/hg38_enb/   \n",
       "caudateNucleus_v2_panpan             /public/home/liuzj/data/singleCellRef/panpan/   \n",
       "caudateNucleus_v2_panTro             /public/home/liuzj/data/singleCellRef/panTro/   \n",
       "caudateNucleus_v2_mmul              /public/home/liuzj/data/singleCellRef/mmul_10/   \n",
       "cerebellarGrayMatter_v2_hg         /public/home/liuzj/data/singleCellRef/hg38_enb/   \n",
       "cerebellarGrayMatter_v2_panpan       /public/home/liuzj/data/singleCellRef/panpan/   \n",
       "cerebellarGrayMatter_v2_panTro       /public/home/liuzj/data/singleCellRef/panTro/   \n",
       "cerebellarGrayMatter_v2_mmul        /public/home/liuzj/data/singleCellRef/mmul_10/   \n",
       "\n",
       "                                                                  id  \\\n",
       "anteriorCingulateCortex_v2_hg          anteriorCingulateCortex_v2_hg   \n",
       "anteriorCingulateCortex_v2_panpan  anteriorCingulateCortex_v2_panpan   \n",
       "anteriorCingulateCortex_v2_panTro  anteriorCingulateCortex_v2_panTro   \n",
       "anteriorCingulateCortex_v2_mmul      anteriorCingulateCortex_v2_mmul   \n",
       "caudateNucleus_v2_hg                            caudateNucleus_v2_hg   \n",
       "caudateNucleus_v2_panpan                    caudateNucleus_v2_panpan   \n",
       "caudateNucleus_v2_panTro                    caudateNucleus_v2_panTro   \n",
       "caudateNucleus_v2_mmul                        caudateNucleus_v2_mmul   \n",
       "cerebellarGrayMatter_v2_hg                cerebellarGrayMatter_v2_hg   \n",
       "cerebellarGrayMatter_v2_panpan        cerebellarGrayMatter_v2_panpan   \n",
       "cerebellarGrayMatter_v2_panTro        cerebellarGrayMatter_v2_panTro   \n",
       "cerebellarGrayMatter_v2_mmul            cerebellarGrayMatter_v2_mmul   \n",
       "\n",
       "                                                              outDir  \n",
       "anteriorCingulateCortex_v2_hg          anteriorCingulateCortex_v2_hg  \n",
       "anteriorCingulateCortex_v2_panpan  anteriorCingulateCortex_v2_panpan  \n",
       "anteriorCingulateCortex_v2_panTro  anteriorCingulateCortex_v2_panTro  \n",
       "anteriorCingulateCortex_v2_mmul      anteriorCingulateCortex_v2_mmul  \n",
       "caudateNucleus_v2_hg                            caudateNucleus_v2_hg  \n",
       "caudateNucleus_v2_panpan                    caudateNucleus_v2_panpan  \n",
       "caudateNucleus_v2_panTro                    caudateNucleus_v2_panTro  \n",
       "caudateNucleus_v2_mmul                        caudateNucleus_v2_mmul  \n",
       "cerebellarGrayMatter_v2_hg                cerebellarGrayMatter_v2_hg  \n",
       "cerebellarGrayMatter_v2_panpan        cerebellarGrayMatter_v2_panpan  \n",
       "cerebellarGrayMatter_v2_panTro        cerebellarGrayMatter_v2_panTro  \n",
       "cerebellarGrayMatter_v2_mmul            cerebellarGrayMatter_v2_mmul  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_runCellRanger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fa701e26",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:57:37.043780Z",
     "start_time": "2021-07-12T10:57:37.033011Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-12 18:57:37.036 | INFO     | jpy_tools.parseSnake2:addRule:55 - runCellRanger step num: 2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "df_runCellRanger = pd.DataFrame(config['samples']).T\n",
       "\n",
       "df_runCellRanger = df_runCellRanger.explode(\"runCellRanger\")\n",
       "df_runCellRanger.index = df_runCellRanger.index + \"_\" + df_runCellRanger.runCellRanger\n",
       "df_runCellRanger['cellRangerIndex'] = df_runCellRanger['runCellRanger'].map(config['dt_genome_cellRangerIndex'])\n",
       "df_runCellRanger['id'] = df_runCellRanger.index\n",
       "df_runCellRanger['outDir'] = df_runCellRanger.index\n",
       "sample_specieLs = df_runCellRanger.index.to_list()\n",
       "for column in ['outDir']:\n",
       "    df_runCellRanger[column] = resultDir + 'step2_runCellRanger/' + df_runCellRanger[column]\n",
       "----------------\n",
       "IN RULE\n",
       "----------------\n",
       "rule runCellRanger:\n",
       "    input:\n",
       "        sampleDir = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'sampleDir'],\n",
       "    output:\n",
       "        runCellRangerFinished = resultDir + 'step2_runCellRanger/' + '{sample_specie}.finished',\n",
       "    params:\n",
       "        gpu = 0,\n",
       "        includeIntron = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'includeIntron'],\n",
       "        expectedCell = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'expectedCell'],\n",
       "        r2Length = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'r2Length'],\n",
       "        id = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'id'],\n",
       "        outDir = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'outDir'],\n",
       "        cellRangerIndex = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'cellRangerIndex'],\n",
       "    threads:48\n",
       "    priority:0\n",
       "    shell:\n",
       "        \"\"\"\n",
       "mkdir -p {params.outDir}\n",
       "cd {params.outDir}\n",
       "if [ {params.includeIntron} = True ]\n",
       "then\n",
       "    cellranger6 count --id {params.id} --include-introns --transcriptome {params.cellRangerIndex} --fastqs {input.sampleDir} --expect-cells {params.expectedCell} --localcores {threads} --r2-length {params.r2Length}\n",
       "else\n",
       "    cellranger6 count --id {params.id} --transcriptome {params.cellRangerIndex} --fastqs {input.sampleDir} --expect-cells {params.expectedCell} --localcores {threads} --r2-length {params.r2Length}\n",
       "fi\n",
       "touch {output.runCellRangerFinished}\n",
       "        \"\"\""
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "runCellRanger = jps.SnakeRule(sf, 'runCellRanger', 48, outFile = '{sample_specie}.finished',)\n",
    "runCellRanger.addCode(\"\"\"\n",
    "df_runCellRanger = pd.DataFrame(config['samples']).T\n",
    "\n",
    "df_runCellRanger = df_runCellRanger.explode(\"runCellRanger\")\n",
    "df_runCellRanger.index = df_runCellRanger.index + \"_\" + df_runCellRanger.runCellRanger\n",
    "df_runCellRanger['cellRangerIndex'] = df_runCellRanger['runCellRanger'].map(config['dt_genome_cellRangerIndex'])\n",
    "df_runCellRanger['id'] = df_runCellRanger.index\n",
    "df_runCellRanger['outDir'] = df_runCellRanger.index\n",
    "sample_specieLs = df_runCellRanger.index.to_list()\n",
    "\"\"\")\n",
    "runCellRanger.addMetaDf('df_runCellRanger', ['outDir'])\n",
    "runCellRanger.addMain('input', ['sampleDir'], wildCardName='sample_specie')\n",
    "runCellRanger.addMain(\"params\", ['includeIntron', 'expectedCell', 'r2Length', 'id', 'outDir','cellRangerIndex'], wildCardName='sample_specie')\n",
    "runCellRanger.setShell(\"\"\"\n",
    "mkdir -p {params.outDir}\n",
    "cd {params.outDir}\n",
    "if [ {params.includeIntron} = True ]\n",
    "then\n",
    "    cellranger6 count --id {params.id} --include-introns --transcriptome {params.cellRangerIndex} --fastqs {input.sampleDir} --expect-cells {params.expectedCell} --localcores {threads} --r2-length {params.r2Length}\n",
    "else\n",
    "    cellranger6 count --id {params.id} --transcriptome {params.cellRangerIndex} --fastqs {input.sampleDir} --expect-cells {params.expectedCell} --localcores {threads} --r2-length {params.r2Length}\n",
    "fi\n",
    "\"\"\")\n",
    "runCellRanger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "be31194c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:57:37.050863Z",
     "start_time": "2021-07-12T10:57:37.045643Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rule all:\n",
       "    input:\n",
       "        runStarsoloFinished = [resultDir + 'step1_runStarsolo/' + \"\" + sample + \".finished\" for sample in sampleLs],\n",
       "        runCellRangerFinished = [resultDir + 'step2_runCellRanger/' + \"\" + sample_specie + \".finished\" for sample_specie in sample_specieLs],"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snakeAll = jps.SnakeAll(sf, runStarsolo, runCellRanger)\n",
    "snakeAll"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "31b085cc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-07-12T10:57:37.057331Z",
     "start_time": "2021-07-12T10:57:37.052747Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "import pandas as pd\n",
      "#configfile: \"/public/home/liuzj/scripts/pipeline/10xAnalysis_concatenatedGenome/snakemake/config.yaml\"\n",
      "pipelineDir = config['pipelineDir']\n",
      "resultDir = config[\"resultDir\"].rstrip(\"/\") + \"/\"\n",
      "pipelineDir = config[\"pipelineDir\"].rstrip(\"/\") + \"/\"\n",
      "sampleLs = list(config['samples'].keys())\n",
      "import glob\n",
      "df_runStarsolo = pd.DataFrame(config['samples']).T\n",
      "df_runStarsolo = df_runStarsolo.assign(starIndex = config['concatenatedStarIndex'], outputDir=df_runStarsolo.index + '/')\n",
      "df_runStarsolo['sampleR2'] = df_runStarsolo['sampleDir'].map(lambda x:','.join(glob.glob(f\"{x}/*/*R2*.fastq.gz\")))\n",
      "df_runStarsolo['sampleR1'] = df_runStarsolo['sampleDir'].map(lambda x:','.join(glob.glob(f\"{x}/*/*R1*.fastq.gz\")))\n",
      "for column in ['outputDir']:\n",
      "    df_runStarsolo[column] = resultDir + 'step1_runStarsolo/' + df_runStarsolo[column]\n",
      "\n",
      "df_runCellRanger = pd.DataFrame(config['samples']).T\n",
      "\n",
      "df_runCellRanger = df_runCellRanger.explode(\"runCellRanger\")\n",
      "df_runCellRanger.index = df_runCellRanger.index + \"_\" + df_runCellRanger.runCellRanger\n",
      "df_runCellRanger['cellRangerIndex'] = df_runCellRanger['runCellRanger'].map(config['dt_genome_cellRangerIndex'])\n",
      "df_runCellRanger['id'] = df_runCellRanger.index\n",
      "df_runCellRanger['outDir'] = df_runCellRanger.index\n",
      "sample_specieLs = df_runCellRanger.index.to_list()\n",
      "for column in ['outDir']:\n",
      "    df_runCellRanger[column] = resultDir + 'step2_runCellRanger/' + df_runCellRanger[column]\n",
      "\n",
      "rule all:\n",
      "    input:\n",
      "        runStarsoloFinished = [resultDir + 'step1_runStarsolo/' + \"\" + sample + \".finished\" for sample in sampleLs],\n",
      "        runCellRangerFinished = [resultDir + 'step2_runCellRanger/' + \"\" + sample_specie + \".finished\" for sample_specie in sample_specieLs],\n",
      "\n",
      "rule runStarsolo:\n",
      "    input:\n",
      "        sampleDir = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'sampleDir'],\n",
      "    output:\n",
      "        runStarsoloFinished = resultDir + 'step1_runStarsolo/' + '{sample}.finished',\n",
      "    params:\n",
      "        gpu = 0,\n",
      "        expectedCell = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'expectedCell'],\n",
      "        umiLength = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'umiLength'],\n",
      "        whiteList = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'whiteList'],\n",
      "        includeIntron = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'includeIntron'],\n",
      "        starIndex = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'starIndex'],\n",
      "        outputDir = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'outputDir'],\n",
      "        sampleR2 = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'sampleR2'],\n",
      "        sampleR1 = lambda wildcard: df_runStarsolo.at[wildcard.sample, 'sampleR1'],\n",
      "    threads:32\n",
      "    priority:20\n",
      "    shell:\n",
      "        \"\"\"\n",
      "mkdir -p {params.outputDir}\n",
      "cd {params.outputDir}\n",
      "if [ {params.includeIntron} = True ]\n",
      "then\n",
      "    STAR --runThreadN {threads} --genomeDir {params.starIndex} --soloType CB_UMI_Simple --soloCBwhitelist {params.whiteList}        --soloUMIlen {params.umiLength} --readFilesIn {params.sampleR2} {params.sampleR1} --soloFeatures Gene GeneFull Velocyto        --clipAdapterType CellRanger4 --outFilterScoreMin 30 --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --soloUMIfiltering MultiGeneUMI_CR        --soloUMIdedup 1MM_CR --soloCellFilter  EmptyDrops_CR {params.expectedCell} 0.99 10 45000 90000 500 0.01 20000 0.01 10000         --readFilesCommand zcat\n",
      "else\n",
      "    STAR --runThreadN {threads} --genomeDir {params.starIndex} --soloType CB_UMI_Simple --soloCBwhitelist {params.whiteList}        --soloUMIlen {params.umiLength} --readFilesIn {params.sampleR2} {params.sampleR1} --soloFeatures Gene Velocyto        --clipAdapterType CellRanger4 --outFilterScoreMin 30 --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts --soloUMIfiltering MultiGeneUMI_CR        --soloUMIdedup 1MM_CR --soloCellFilter  EmptyDrops_CR {params.expectedCell} 0.99 10 45000 90000 500 0.01 20000 0.01 10000         --readFilesCommand zcat\n",
      "fi\n",
      "samtools view -@ {threads} -S -O BAM ./Aligned.out.sam  | samtools sort -@ {threads} - -O BAM -o ./Aligned.out.bam\n",
      "samtools index -@ {threads} ./Aligned.out.bam\n",
      "rm ./Aligned.out.sam\n",
      "touch {output.runStarsoloFinished}\n",
      "        \"\"\"\n",
      "\n",
      "rule runCellRanger:\n",
      "    input:\n",
      "        sampleDir = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'sampleDir'],\n",
      "    output:\n",
      "        runCellRangerFinished = resultDir + 'step2_runCellRanger/' + '{sample_specie}.finished',\n",
      "    params:\n",
      "        gpu = 0,\n",
      "        includeIntron = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'includeIntron'],\n",
      "        expectedCell = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'expectedCell'],\n",
      "        r2Length = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'r2Length'],\n",
      "        id = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'id'],\n",
      "        outDir = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'outDir'],\n",
      "        cellRangerIndex = lambda wildcard: df_runCellRanger.at[wildcard.sample_specie, 'cellRangerIndex'],\n",
      "    threads:48\n",
      "    priority:0\n",
      "    shell:\n",
      "        \"\"\"\n",
      "mkdir -p {params.outDir}\n",
      "cd {params.outDir}\n",
      "if [ {params.includeIntron} = True ]\n",
      "then\n",
      "    cellranger6 count --id {params.id} --include-introns --transcriptome {params.cellRangerIndex} --fastqs {input.sampleDir} --expect-cells {params.expectedCell} --localcores {threads} --r2-length {params.r2Length}\n",
      "else\n",
      "    cellranger6 count --id {params.id} --transcriptome {params.cellRangerIndex} --fastqs {input.sampleDir} --expect-cells {params.expectedCell} --localcores {threads} --r2-length {params.r2Length}\n",
      "fi\n",
      "touch {output.runCellRangerFinished}\n",
      "        \"\"\"\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sf.getMain(path_snakeFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b18261d5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
